crates/ibc/src/clients/ics07_tendermint/host_helpers.rs

```diff
@@ -1,131 +1,122 @@
-use core::time::Duration;
-
 use alloc::format;
-use alloc::string::{String, ToString};
+use alloc::string::ToString;
 use ibc_proto::google::protobuf::Any;
 
 use crate::clients::ics07_tendermint::client_state::ClientState as TmClientState;
+use crate::core::context::HostChainContext;
 use crate::core::ics02_client::client_state::ClientState;
-use crate::core::ics03_connection::error::ConnectionError;
-use crate::core::ics23_commitment::specs::ProofSpecs;
-use crate::core::ics24_host::identifier::ChainId;
-use crate::Height;
+use crate::core::ics24_host::error::HostError;
 
 use tendermint::trust_threshold::TrustThresholdFraction as TendermintTrustThresholdFraction;
 
-/// Provides an implementation of `ValidationContext::validate_self_client` for
-/// Tendermint-based hosts.
-pub trait ValidateSelfClientContext {
-    fn validate_self_client(&self, counterparty_client_state: Any) -> Result<(), ConnectionError> {
-        let counterparty_client_state = TmClientState::try_from(counterparty_client_state)
-            .map_err(|_| ConnectionError::InvalidClientState {
+/// Provides an implementation of `validate_self_client` for Tendermint-based
+/// hosts.
+pub fn validate_self_tendermint_client<Ctx>(
+    ctx: &Ctx,
+    counterparty_client_state: Any,
+) -> Result<(), HostError>
+where
+    Ctx: HostChainContext,
```

Comment:
`ValidateSelfClientContext` is designed to ask for an implementation of all the methods that it needs in its implementation; some only implementable by Tendermint hosts (such as `unbonding_period()`). I believe this is the root cause for why you had to change the signature to `Option<Duration>`; not all hosts can implement it.

This is why I don't think it is appropriate for `ValidationContext` to have `HostChainContext` as a supertrait. Hypothetically, say we also provided the implementation of `validate_self_client()` for NEAR hosts; would we need to have that `Context` be a supertrait of `ValidationContext` as well, requiring all hosts to implement NEAR-specific methods (and have the signature return an `Option` for those too)?

Reply:
I understand your concern, but rather than the `unbonding_period()` method, which I also found it there to be a wart, other methods should be available by any chain